File: Microsoft/Array/54. Spiral Matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        re = []
        if not matrix:
            return re
        
        r1, r2 = 0, len(matrix) - 1
        c1, c2 = 0, len(matrix[0]) - 1
        
        while r1 <= r2 and c1 <= c2:
            for c in range(c1, c2+1):
                re.append(matrix[r1][c])
            for r in range(r1+1, r2+1):
                re.append(matrix[r][c2])
            if r1 < r2 and c1 < c2:
                for c in range(c2-1, c1-1, -1):
                    re.append(matrix[r2][c])
                for r in range(r2-1, r1, -1):
                    re.append(matrix[r][c1])
                
            r1 += 1
            r2 -= 1
            c1 += 1
            c2 -= 1
            
        return re
        

File: Top Interview Questions/Array/54. Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> re = new ArrayList<Integer>();
        
        if (matrix.length == 0) return re;
        
        int lr = 0, rr = matrix.length - 1;
        int lc = 0, rc = matrix[0].length - 1;
        
        while (lr <= rr && lc <= rc) {
            for (int i = lc; i <= rc; i++) {
                re.add(matrix[lr][i]);
            }
            
            for (int i = lr + 1; i <= rr; i++) {
                re.add(matrix[i][rc]);
            }
            
            if (lr < rr && lc < rc){
                for (int i = rc - 1; i >= lc; i--) {
                    re.add(matrix[rr][i]);
                }

                for (int i = rr - 1; i > lr; i--) {
                    re.add(matrix[i][lc]);
                }
            }
            lr++;
            rr--;
            lc++;
            rc--;
        }
        
        return re;
    }
}